audit/hashstructure.go

```diff
@@ -19,107 +19,156 @@ func HashString(salter *salt.Salt, data string) string {
+// recognized, then it will be passed through unchanged.
+func HashAuth(salter *salt.Salt, in *logical.Auth, HMACAccessor bool) (*logical.Auth, error) {
+	if in == nil {
+		return nil, nil
+	}
+
 	fn := salter.GetIdentifiedHMAC
+	auth := *in
 
-	switch s := raw.(type) {
-	case *logical.Auth:
-		if s == nil {
-			return nil
-		}
-		if s.ClientToken != "" {
-			s.ClientToken = fn(s.ClientToken)
-		}
-		if s.Accessor != "" {
-			s.Accessor = fn(s.Accessor)
-		}
+	if auth.ClientToken != "" {
+		auth.ClientToken = fn(auth.ClientToken)
+	}
+	if HMACAccessor && auth.Accessor != "" {
+		auth.Accessor = fn(auth.Accessor)
+	}
+	return &auth, nil
+}
 
-	case *logical.Request:
-		if s == nil {
-			return nil
-		}
-		if s.Auth != nil {
-			if err := Hash(salter, s.Auth, nil); err != nil {
-				return err
-			}
-		}
+func HashRequest(salter *salt.Salt, in *logical.Request, HMACAccessor bool, nonHMACDataKeys []string) (*logical.Request, error) {
+	if in == nil {
+		return nil, nil
+	}
 
-		if s.ClientToken != "" {
-			s.ClientToken = fn(s.ClientToken)
-		}
+	fn := salter.GetIdentifiedHMAC
+	req := *in
 
-		if s.ClientTokenAccessor != "" {
-			s.ClientTokenAccessor = fn(s.ClientTokenAccessor)
+	if req.Auth != nil {
+		cp, err := copystructure.Copy(req.Auth)
+		if err != nil {
+			return nil, err
 		}
 
-		data, err := HashStructure(s.Data, fn, nonHMACDataKeys)
+		req.Auth, err = HashAuth(salter, cp.(*logical.Auth), HMACAccessor)
 		if err != nil {
-			return err
+			return nil, err
 		}
+	}
 
-		s.Data = data.(map[string]interface{})
+	if req.ClientToken != "" {
+		req.ClientToken = fn(req.ClientToken)
+	}
+	if HMACAccessor && req.ClientTokenAccessor != "" {
+		req.ClientTokenAccessor = fn(req.ClientTokenAccessor)
+	}
 
-	case *logical.Response:
-		if s == nil {
-			return nil
-		}
+	data, err := hashMap(fn, req.Data, nonHMACDataKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Data = data
+	return &req, nil
+}
+
+func hashMap(fn func(string) string, data map[string]interface{}, nonHMACDataKeys []string) (map[string]interface{}, error) {
+	if data == nil {
+		return nil, nil
+	}
 
-		if s.Auth != nil {
-			if err := Hash(salter, s.Auth, nil); err != nil {
-				return err
+	copy, err := copystructure.Copy(data)
+	if err != nil {
+		return nil, err
+	}
+	newData := copy.(map[string]interface{})
+	for k, v := range newData {
+		if o, ok := v.(logical.OptMarshaler); ok {
+			marshaled, err := o.MarshalJSONWithOptions(&logical.MarshalOptions{
+				ValueHasher: fn,
+			})
+			if err != nil {
+				return nil, err
 			}
+			newData[k] = json.RawMessage(marshaled)
 		}
+	}
 
-		if s.WrapInfo != nil {
-			if err := Hash(salter, s.WrapInfo, nil); err != nil {
-				return err
-			}
+	if err := HashStructure(newData, fn, nonHMACDataKeys); err != nil {
+		return nil, err
+	}
+
+	return newData, nil
+}
+
+func HashResponse(salter *salt.Salt, in *logical.Response, HMACAccessor bool, nonHMACDataKeys []string) (*logical.Response, error) {
```

Comment:
In looking at the callers, I'm wondering if it might make sense to be extra defensive about the `salter` and to add a guard for a nil pointer there. It _should_ never be nil, but it's pulled out of an interface that can be backed by multiple structs, and the interface doesn't state that part of its contract is that an error and a nil pointer are mutually exclusive from the `Salt` method.

Reply:
That's my thought for the other methods here that take the `salter`, as well.

Reply:
I'm flexible on that, though, I do see that it would be inconvenient because `HashString` would then need to return an err.

Reply:
We chatted about this and agreed that it was ok to assume the pointer would be fulfilled, and might also add a comment to the `Salt` interface regarding an error and pointer being mutually exclusive.